File: cpp/day11.cpp
// day11.cpp
#include <gtest/gtest.h>
#include <string>
#include <fstream>
#include <numeric>
using namespace std;

class Day11 {
protected:
    int grid[10][10];
    set<tuple<int,int>> flashed;

    void inc(int r, int c) {
        if (grid[r][c] > 9) return;
        grid[r][c]++;
        if (grid[r][c] > 9) {
            flashed.insert(make_tuple(r,c));
            for (int dr=r-1; dr<r+2; dr++) {
                if (dr >= 0 && dr <= 9) {
                    for (int dc=c-1; dc<c+2; dc++) {
                        if (dc >=0 && dc <=9) {
                            inc(dr, dc);
                        }
                    }
                }
            }
        }
    }

    void reset(const tuple<int,int>& t) {
        grid[get<0>(t)][get<1>(t)] = 0;
    }

    void do_step() {
        flashed.clear();
        for (int row=0; row<10; row++) {
            for (int col=0; col<10; col++) {
                inc(row, col);
            }
        }
        //for_each(flashed.cbegin(), flashed.cend(), [this](const tuple<int,int> &t){reset(t);});
        for (auto &t : flashed) { reset(t); }
    }

public:
    Day11(const char *file) {
        ifstream infile(file);
        string line;
        int row = 0;
        while (getline(infile,line)) {
            for (int col=0; col<10; col++) {
                grid[row][col] = line[col]-'0';
            }
            row++;
        }
        infile.close();
    }

    int part1() {
        int count = 0;
        for (int i=0; i<100; i++) {
            do_step();
            count += flashed.size();
        }
        return count;
    }

    int part2() {
        for (int i=0; i<1000; i++) {
            do_step();
            if (flashed.size() == 100) {
                return i+1;
            }
        }
        return 0;
    }
};

int part2(ifstream &f) {
    return 0;
}

TEST(Day11, Part1_test) {
    ASSERT_EQ(Day11("../test11.txt").part1(), 1656);

}

TEST(Day11, Part1) {
    ASSERT_EQ(Day11("../input11.txt").part1(), 1741);

}

TEST(Day11, Part2_test) {
    ASSERT_EQ(Day11("../test11.txt").part2(), 195);

}

TEST(Day11, Part2) {
    ASSERT_EQ(Day11("../input11.txt").part2(), 440);

}

File: day11.py
#day11.py 2021
import unittest

def delta(r,c):
    for dr in range(r-1,r+2):
        if dr >=0 and dr <=9:
            for dc in range(c-1,c+2):
                if dc >=0 and dc <=9:
                    yield(dr,dc)

def do_step(grid, flashed, r, c):
    if grid[r][c] > 9:
        return
    grid[r][c] += 1
    if grid[r][c] > 9:
        flashed.add((r,c))
        for dr,dc in delta(r,c):
            do_step(grid, flashed, dr, dc)

def part1(lines):
    grid = []
    count = 0
    for line in lines:
        grid.append([int(n) for n in line.strip()])
    for step in range(100):
        flashed = set()
        for r,row in enumerate(grid):
            for c,octopus in enumerate(row):
                do_step(grid, flashed, r, c)
        for r,c in flashed:
            grid[r][c] = 0
        count += len(flashed)
    return count

def part2(lines):
    grid = []
    for line in lines:
        grid.append([int(n) for n in line.strip()])
    for step in range(1000):
        flashed = set()
        for r,row in enumerate(grid):
            for c,octopus in enumerate(row):
                do_step(grid, flashed, r, c)
        for r,c in flashed:
            grid[r][c] = 0
        if len(flashed) == 100:
            return step+1
    return None

class TestDay11(unittest.TestCase):
    def test_1a(self):
        with open('./test11.txt', 'r') as f:
            self.assertEqual(part1(list(f)), 1656)

    def test_1(self):
        with open('./input11.txt', 'r') as f:
            self.assertEqual(part1(list(f)), 1741)

    def test_2a(self):
        with open('./test11.txt', 'r') as f:
            self.assertEqual(part2(list(f)), 195)

    def test_2(self):
        with open('./input11.txt', 'r') as f:
            self.assertEqual(part2(list(f)), 440)

if __name__ == '__main__':
    unittest.main()
